tokio/src/io/driver/mod.rs

```diff
@@ -102,16 +121,27 @@ impl Driver {
     }
 
     fn turn(&mut self, max_wait: Option<Duration>) -> io::Result<()> {
+        // How often to call `compact()` on the resource slab
+        const COMPACT_INTERVAL: u16 = 256;
+
+        self.tick = self.tick.wrapping_add(1);
+
+        if self.tick % COMPACT_INTERVAL == 0 {
+            self.resources.compact();
+        }
```

Comment:
It seems like the kind of brute force approach of doing a compaction pass every _n_ ticks could lead to spurious compactions, and thus unnecessarily locking the slab. 

Would it be possible to do this based on counting the number of resources deregistered, or something? That way, we would only do a compaction when we know that there is a diff in the number of resources, avoiding the overhead of compacting if no resources have been deallocated since the last compaction.

Reply:
I added a hint in `compact()` that avoids locking if there is no work to be done.